proposals/0000-deep-subsumption.rst

```diff
@@ -0,0 +1,304 @@
+
+Like any other extension, ``DeepSubsumption`` can be turned on or off with a ``LANGUAGE`` pragma
+or in a ``.cabal`` file. Because ``DeepSubsumption`` is part of ``Haskell2010`` and ``Haskell98``,
+projects compiled with ``.cabal`` files that declare either of these to be the ``default-language``
+will get the benefits of deep subsumption, much like it was implemented prior to GHC-9.0.
+Note that any project using ``GHC2021`` must also have upgraded to use shallow subsumption, because
+the two features came out in the same GHC release.
+
+2.1 DeepSubsumption is not recommended
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``DeepSubsumption`` extension is not recommended:
+
+* It makes the runtime semantics (including performance) of Haskell programs
+  less predictable (due to silent eta-expansion), as the original proposal describes.
+  The situation is even more complicated when type classes are involved.  You can find some intricate discussion on the `Simplified subsumption proposal discussion thread <https://github.com/ghc-proposals/ghc-proposals/pull/287>`_, especially towards the end.
+
+* The interaction between ``DeepSubsumption`` and ``ImpredicativeTypes`` is hard to predict.  Quick Look treats function arrow as invariant, which is different to ``DeepSubsumption``, but it is hard to come up with concrete examples that show strange behaviour.  Perhaps surprisingly, the two different treatments of function arrow, while infelicitous, do not seem to have an immediately bad effects.
+
+* ``DeepSubsumption`` (notably deep skolemisation) seems to be fundamentally incompatible with the accepted proposal 155: `Binding type variables in lambda expressions <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0155-type-lambda.rst>`_. Consider::
+
+      f :: Int -> forall a. a -> a
+      f = \x-> let v = x+x in
+               \ @a -> blah
+
+  To correctly bind the ``\ @a`` we must not deeply skolemise ``f``'s type at the outset;
+  yet we must do so to implement ``DeepSubsumption``.  So GHC will reject an attempt to switch both on at once.
+
+Despite these shortcomings, in a manner similar to
+``NoMonoLocalBinds``, users who really want such a feature are free to
+enable ``DeepSubsumption``, with the understanding that doing so might
+introduce changes to type inference or runtime behaviour that are
+difficult to predict.
+
+Despite not being recommended, there is no deprecation plan for ``DeepSubsumption``, it will
+be available as an extension indefinitely.
+
+2.2 Warnings
+^^^^^^^^^^^^
+
+Given that we don't think that using ``DeepSubsumption`` is a good idea, we also
+propose to improve diagnostics to help users migrate to simplified
+subsumption.
+
+* When ``-XDeepSubsumption`` is off, the error message can be improved to suggest
+  eta-expansion (and optionally enabling ``DeepSubsumption``).
+
+* When ``-XDeepSubsumption`` is on:
+
+  * A new warning ``-Wdeep-subsumption``, in ``-Wcompat`` and ``-Weverything`` (but
```

Comment:
The [User Guide](https://downloads.haskell.org/ghc/latest/docs/html/users_guide/using-warnings.html?highlight=wcompat#ghc-flag--Wcompat) says that `-Wcompat` is for warnings that will be enabled by default in the future. Is that the plan for `-Wdeep-subsumption`? If so, over what timeframe?

I also think it would be weird if we landed in a world where a language extension enabled a feature where all uses incur a warning. This proposal in particular seems to be motivated by pain felt by industrial users, who also tend to set `-Werror` to keep their code warning-free. Are we not just going to cause the same pain via the warnings?

Reply:
I believe that `-XPartialTypeSignatures` is already a GHC extension that enables a feature, all of whose uses incur a warning.  That said, I agree that it's weird, in that case as well as this one.